File: api/app/domains/compliance/compliance_calc.py
"""Compliance-calendar logic — pure, deterministic. Operates on plain entry dicts
({domain, form_name, due_date, status}) so it is trivially testable. Time is injected
via ``as_of``.
"""

from __future__ import annotations

from datetime import date, timedelta
from typing import Any

# Statutory domains that map 1:1 onto the compliance sub-vector dimensions.
STATUTORY_DOMAINS = ("roc", "gst", "tds", "pf", "esi", "pt")


def mca_deadlines(*, agm_date: str) -> list[dict[str, Any]]:
    """MCA annual-filing deadlines for a private company given its AGM date (Companies Act
    2013): AOC-4 within 30 days, MGT-7 within 60; DPT-3 by 30 Jun and DIR-3 KYC by 30 Sep."""
    agm = date.fromisoformat(agm_date)
    return [
        {"domain": "roc", "form": "AOC-4", "statute": "Companies Act 2013 s.137",
         "due_date": (agm + timedelta(days=30)).isoformat()},
        {"domain": "roc", "form": "MGT-7", "statute": "Companies Act 2013 s.92",
         "due_date": (agm + timedelta(days=60)).isoformat()},
        {"domain": "roc", "form": "DPT-3", "statute": "Companies Act 2013 Rule 16",
         "due_date": f"{agm.year}-06-30"},
        {"domain": "roc", "form": "DIR-3 KYC", "statute": "Companies Act 2013 Rule 12A",
         "due_date": f"{agm.year}-09-30"},
    ]

def _add_months(d: date, months: int) -> date:
    total = d.month - 1 + months
    year = d.year + total // 12
    month = total % 12 + 1
    import calendar

    day = min(d.day, calendar.monthrange(year, month)[1])
    return date(year, month, day)


# ---- Secretarial compliance (minutes / AGM / resolutions) -----------------------------


def board_meeting_compliance(meeting_dates: list[str]) -> dict[str, Any]:
    """Companies Act 2013 s.173: a company needs >=4 board meetings a year, with no gap
    exceeding 120 days between two consecutive meetings. Pure."""
    dates = sorted(date.fromisoformat(d) for d in meeting_dates)
    count_ok = len(dates) >= 4
    gaps = [(dates[i] - dates[i - 1]).days for i in range(1, len(dates))]
    max_gap = max(gaps) if gaps else 0
    gap_ok = max_gap <= 120
    reasons = []
    if not count_ok:
        reasons.append("fewer than 4 board meetings in the year")
    if len(dates) > 1 and not gap_ok:
        reasons.append(f"gap of {max_gap} days exceeds 120 between consecutive meetings")
    return {
        "compliant": count_ok and gap_ok,
        "meetings": len(dates),
        "max_gap_days": max_gap,
        "reasons": reasons,
    }


def secretarial_calendar(fy_end: str) -> list[dict[str, Any]]:
    """Annual secretarial calendar for a private company given its FY end (YYYY-MM-DD): the AGM
    (within 6 months of FY end, s.96) plus the recurring board-meeting, register and minutes
    duties spread across the year. Pure."""
    end = date.fromisoformat(fy_end)
    fy_start = _add_months(end, -12)
    items = [
        {"item": "Annual General Meeting", "statute": "Companies Act 2013 s.96",
         "due_date": _add_months(end, 6).isoformat()},
        {"item": "Maintain statutory registers", "statute": "Companies Act 2013 s.88",
         "due_date": end.isoformat()},
        {"item": "Record minutes of meetings", "statute": "Companies Act 2013 s.118",
         "due_date": end.isoformat()},
    ]
    for q in range(4):  # one board meeting target per quarter
        items.append(
            {"item": f"Board meeting Q{q + 1}", "statute": "Companies Act 2013 s.173",
             "due_date": _add_months(fy_start, q * 3 + 3).isoformat()}
        )
    return items


# ---- Audit support package ------------------------------------------------------------

AUDIT_CHECKLIST = (
    "trial_balance", "general_ledger", "bank_statements", "bank_reconciliation",
    "fixed_asset_register", "gst_returns", "tds_returns", "payroll_records",
    "invoices", "bills", "board_minutes", "cap_table",
)


def audit_support_package(available: set[str], *, audit_type: str = "statutory") -> dict[str, Any]:
    """Audit support checklist (statutory or internal): which schedules/records are ready and
    which are missing, with a readiness percentage. Pure."""
    have = set(available)
    missing = [item for item in AUDIT_CHECKLIST if item not in have]
    present = len(AUDIT_CHECKLIST) - len(missing)
    return {
        "audit_type": audit_type,
        "items": [{"item": i, "present": i in have} for i in AUDIT_CHECKLIST],
        "missing": missing,
        "readiness_pct": round(100.0 * present / len(AUDIT_CHECKLIST), 1),
    }


# ---- DPIIT Startup India recognition --------------------------------------------------

_DPIIT_MAX_TURNOVER_PAISE = 100 * 10**7 * 100  # Rs 100 crore, in paise


def dpiit_eligibility(
    *,
    incorporation_date: str,
    as_of: str,
    annual_turnover_paise: int,
    is_reconstituted: bool = False,
) -> dict[str, Any]:
    """DPIIT Startup India recognition (Notification G.S.R. 127(E)): incorporated < 10 years
    ago, annual turnover < Rs.100 crore, and not formed by splitting/reconstructing an existing
    business. Pure & deterministic (time injected via ``as_of``)."""
    inc = date.fromisoformat(incorporation_date)
    now = date.fromisoformat(as_of)
    age_years = (now - inc).days / 365.25
    age_ok = age_years < 10
    turnover_ok = int(annual_turnover_paise) < _DPIIT_MAX_TURNOVER_PAISE
    reasons = []
    if not age_ok:
        reasons.append("incorporated 10 or more years ago")
    if not turnover_ok:
        reasons.append("annual turnover is Rs.100 crore or more")
    if is_reconstituted:
        reasons.append("formed by splitting up / reconstructing an existing business")
    return {
        "eligible": age_ok and turnover_ok and not is_reconstituted,
        "age_years": round(age_years, 2),
        "reasons": reasons,
    }


# Reminder cadence (PRD §1.10): T-7, T-1, T-0.
ALERT_OFFSETS = {7: "T-7", 1: "T-1", 0: "T-0"}


def _pending(entries: list[dict]) -> list[dict]:
    return [e for e in entries if e.get("status") != "filed"]


def overdue_count(entries: list[dict], as_of: date) -> int:
    return sum(1 for e in _pending(entries) if date.fromisoformat(e["due_date"]) < as_of)


def domain_health(entries: list[dict], as_of: date) -> dict[str, float]:
    """1.0 when a statute has nothing overdue, else 0.0."""
    health = dict.fromkeys(STATUTORY_DOMAINS, 1.0)
    for e in _pending(entries):
        d = e.get("domain")
        if d in health and date.fromisoformat(e["due_date"]) < as_of:
            health[d] = 0.0
    return health


def alerts(entries: list[dict], as_of: date) -> list[dict[str, Any]]:
    """Reminders due exactly on ``as_of`` (T-7/T-1/T-0) plus any already overdue."""
    out: list[dict[str, Any]] = []
    for e in _pending(entries):
        delta = (date.fromisoformat(e["due_date"]) - as_of).days
        if delta < 0:
            out.append({**_label(e), "label": "OVERDUE", "days_overdue": -delta})
        elif delta in ALERT_OFFSETS:
            out.append({**_label(e), "label": ALERT_OFFSETS[delta], "days_to_due": delta})
    return out


def _label(e: dict) -> dict[str, Any]:
    return {"domain": e.get("domain"), "form_name": e.get("form_name"), "due_date": e["due_date"]}

File: api-nest/src/domains/compliance/compliance.calc.ts
/**
 * Compliance-calendar logic — pure, deterministic. Faithful port of
 * api/app/domains/compliance/compliance_calc.py. Time is injected via `asOf`
 * (ISO date string, YYYY-MM-DD). No clock read here.
 *
 * Dates are ISO strings throughout. Day arithmetic uses UTC epoch days to avoid
 * DST drift; ISO date lexicographic order == chronological order for `<` compares.
 */

export const STATUTORY_DOMAINS = ['roc', 'gst', 'tds', 'pf', 'esi', 'pt'] as const;

// ---- date helpers (UTC, no clock) --------------------------------------------

function utcMs(iso: string): number {
  const [y, m, d] = iso.split('-').map((x) => parseInt(x, 10));
  return Date.UTC(y, m - 1, d);
}

/** Whole-day count from `earlier` to `later` (later - earlier), like Python timedelta.days. */
function daysBetween(later: string, earlier: string): number {
  return Math.round((utcMs(later) - utcMs(earlier)) / 86_400_000);
}

function fmt(ms: number): string {
  const d = new Date(ms);
  const y = d.getUTCFullYear();
  const mo = String(d.getUTCMonth() + 1).padStart(2, '0');
  const day = String(d.getUTCDate()).padStart(2, '0');
  return `${y}-${mo}-${day}`;
}

function addDays(iso: string, days: number): string {
  return fmt(utcMs(iso) + days * 86_400_000);
}

// Python floor division / modulo (JS % follows sign of dividend).
function floorDiv(a: number, b: number): number {
  return Math.floor(a / b);
}
function floorMod(a: number, b: number): number {
  return ((a % b) + b) % b;
}
function daysInMonth(year: number, month1: number): number {
  return new Date(Date.UTC(year, month1, 0)).getUTCDate();
}

/** Mirror of _add_months: shift an ISO date by `months`, clamping the day to month length. */
export function addMonths(iso: string, months: number): string {
  const [y, m, d] = iso.split('-').map((x) => parseInt(x, 10));
  const total = m - 1 + months;
  const year = y + floorDiv(total, 12);
  const month = floorMod(total, 12) + 1;
  const day = Math.min(d, daysInMonth(year, month));
  return `${year}-${String(month).padStart(2, '0')}-${String(day).padStart(2, '0')}`;
}

function round(x: number, ndigits: number): number {
  const f = 10 ** ndigits;
  return Math.round(x * f) / f;
}

// ---- MCA annual filings ------------------------------------------------------

export function mcaDeadlines(agmDate: string): Record<string, string>[] {
  const year = agmDate.slice(0, 4);
  return [
    { domain: 'roc', form: 'AOC-4', statute: 'Companies Act 2013 s.137', due_date: addDays(agmDate, 30) },
    { domain: 'roc', form: 'MGT-7', statute: 'Companies Act 2013 s.92', due_date: addDays(agmDate, 60) },
    { domain: 'roc', form: 'DPT-3', statute: 'Companies Act 2013 Rule 16', due_date: `${year}-06-30` },
    { domain: 'roc', form: 'DIR-3 KYC', statute: 'Companies Act 2013 Rule 12A', due_date: `${year}-09-30` },
  ];
}

// ---- Secretarial compliance --------------------------------------------------

export function boardMeetingCompliance(meetingDates: string[]): Record<string, any> {
  const dates = [...meetingDates].sort();
  const countOk = dates.length >= 4;
  const gaps: number[] = [];
  for (let i = 1; i < dates.length; i++) gaps.push(daysBetween(dates[i], dates[i - 1]));
  const maxGap = gaps.length ? Math.max(...gaps) : 0;
  const gapOk = maxGap <= 120;
  const reasons: string[] = [];
  if (!countOk) reasons.push('fewer than 4 board meetings in the year');
  if (dates.length > 1 && !gapOk)
    reasons.push(`gap of ${maxGap} days exceeds 120 between consecutive meetings`);
  return { compliant: countOk && gapOk, meetings: dates.length, max_gap_days: maxGap, reasons };
}

export function secretarialCalendar(fyEnd: string): Record<string, string>[] {
  const fyStart = addMonths(fyEnd, -12);
  const items: Record<string, string>[] = [
    { item: 'Annual General Meeting', statute: 'Companies Act 2013 s.96', due_date: addMonths(fyEnd, 6) },
    { item: 'Maintain statutory registers', statute: 'Companies Act 2013 s.88', due_date: fyEnd },
    { item: 'Record minutes of meetings', statute: 'Companies Act 2013 s.118', due_date: fyEnd },
  ];
  for (let q = 0; q < 4; q++) {
    items.push({
      item: `Board meeting Q${q + 1}`,
      statute: 'Companies Act 2013 s.173',
      due_date: addMonths(fyStart, q * 3 + 3),
    });
  }
  return items;
}

// ---- Audit support package ---------------------------------------------------

export const AUDIT_CHECKLIST = [
  'trial_balance', 'general_ledger', 'bank_statements', 'bank_reconciliation',
  'fixed_asset_register', 'gst_returns', 'tds_returns', 'payroll_records',
  'invoices', 'bills', 'board_minutes', 'cap_table',
] as const;

export function auditSupportPackage(
  available: Iterable<string>,
  auditType = 'statutory',
): Record<string, any> {
  const have = new Set(available);
  const missing = AUDIT_CHECKLIST.filter((i) => !have.has(i));
  const present = AUDIT_CHECKLIST.length - missing.length;
  return {
    audit_type: auditType,
    items: AUDIT_CHECKLIST.map((i) => ({ item: i, present: have.has(i) })),
    missing,
    readiness_pct: round((100.0 * present) / AUDIT_CHECKLIST.length, 1),
  };
}

// ---- DPIIT Startup India recognition -----------------------------------------

const DPIIT_MAX_TURNOVER_PAISE = 100 * 10 ** 7 * 100; // Rs 100 crore, in paise

export function dpiitEligibility(args: {
  incorporationDate: string;
  asOf: string;
  annualTurnoverPaise: number;
  isReconstituted?: boolean;
}): Record<string, any> {
  const isReconstituted = args.isReconstituted ?? false;
  const ageYears = daysBetween(args.asOf, args.incorporationDate) / 365.25;
  const ageOk = ageYears < 10;
  const turnoverOk = Math.trunc(args.annualTurnoverPaise) < DPIIT_MAX_TURNOVER_PAISE;
  const reasons: string[] = [];
  if (!ageOk) reasons.push('incorporated 10 or more years ago');
  if (!turnoverOk) reasons.push('annual turnover is Rs.100 crore or more');
  if (isReconstituted) reasons.push('formed by splitting up / reconstructing an existing business');
  return {
    eligible: ageOk && turnoverOk && !isReconstituted,
    age_years: round(ageYears, 2),
    reasons,
  };
}

// ---- Calendar alerts / health ------------------------------------------------

export type Entry = { domain?: string; form_name?: string; due_date: string; status?: string };

// Reminder cadence (PRD §1.10): T-7, T-1, T-0.
const ALERT_OFFSETS: Record<number, string> = { 7: 'T-7', 1: 'T-1', 0: 'T-0' };

function pending(entries: Entry[]): Entry[] {
  return entries.filter((e) => e.status !== 'filed');
}

export function overdueCount(entries: Entry[], asOf: string): number {
  return pending(entries).filter((e) => e.due_date < asOf).length;
}

export function domainHealth(entries: Entry[], asOf: string): Record<string, number> {
  const health: Record<string, number> = {};
  for (const d of STATUTORY_DOMAINS) health[d] = 1.0;
  for (const e of pending(entries)) {
    if (e.domain && e.domain in health && e.due_date < asOf) health[e.domain] = 0.0;
  }
  return health;
}

function label(e: Entry): Record<string, any> {
  return { domain: e.domain, form_name: e.form_name, due_date: e.due_date };
}

export function alerts(entries: Entry[], asOf: string): Record<string, any>[] {
  const out: Record<string, any>[] = [];
  for (const e of pending(entries)) {
    const delta = daysBetween(e.due_date, asOf);
    if (delta < 0) {
      out.push({ ...label(e), label: 'OVERDUE', days_overdue: -delta });
    } else if (delta in ALERT_OFFSETS) {
      out.push({ ...label(e), label: ALERT_OFFSETS[delta], days_to_due: delta });
    }
  }
  return out;
}
